CRM/MembershipExtras/Service/MembershipTypeInstalmentCalculator.php

```diff
@@ -0,0 +1,248 @@
+   * @var array
+   */
+  private $membershipTypes;
+
+  /**
+   * @var \CRM_Member_BAO_MembershipTypeTaxAmountCalculator
+   */
+  private $membershipTypeTaxAmountCalculator;
+
+  /**
+   * @var string
+   */
+  private $schedule;
+
+  /**
+   * CRM_MembershipExtras_Service_MembershipTypeInstalment constructor.
+   *
+   * @param array $membershipTypes
+   * @param \MembershipTypeTaxAmountCalculator $membershipTypeTaxAmountCalculator
+   * @param \MembershipTypeDatesCalculator $membershipTypeDatesCalculator
+   * @param $schedule
+   * @throws CRM_MembershipExtras_Exception_InvalidMembershipTypeInstalmentCalculator
+   */
+  public function __construct(
+    array $membershipTypes,
+    MembershipTypeTaxAmountCalculator $membershipTypeTaxAmountCalculator,
+    MembershipTypeDatesCalculator $membershipTypeDatesCalculator,
+    $schedule
+  ) {
+    $this->membershipTypes = $membershipTypes;
+    $this->membershipTypeTaxAmountCalculator = $membershipTypeTaxAmountCalculator;
+    $this->membershipTypeDatesCalculator = $membershipTypeDatesCalculator;
+    $this->schedule = $schedule;
+    $this->validateMembershipTypeForInstalment();
+  }
+
+  /**
+   * Calculates the instalment amount for a set of membership types given that the
+   * condition for calculating the following instalment amount is met.
+   * Calculation is (total amount of membership(s) cost for a full year / 12
+   *
+   * @return CRM_MembershipExtras_DTO_ScheduleInstalmentAmount
+   */
+  private function calculateInstalmentAmount() {
+    $totalMembershipTypeAmount = 0;
+    $totalMembershipTypeTaxAmount = 0;
+    foreach ($this->membershipTypes as $membershipType) {
+      $totalMembershipTypeAmount += $membershipType->minimum_fee;
+      $totalMembershipTypeTaxAmount = +$this->membershipTypeTaxAmountCalculator->calculateTax($membershipType);
+    }
+    switch ($this->schedule) {
+      case self::MONTHLY:
+        $divisor = self::MONTHLY_INSTALMENT;
+        break;
+
+      case self::QUARTERLY:
+        $divisor = self::QUARTERLY_INSTALMENT;
+        break;
+
+      default:
+        $divisor = self::ANNUALLY_INSTALMENT;
+    }
+    $instalmentAmount = MoneyUtilities::roundToPrecision($totalMembershipTypeAmount / $divisor, 2);
+    $instalmentTaxAmount = MoneyUtilities::roundToPrecision($totalMembershipTypeTaxAmount / $divisor, 2);
+
+    $instalment = new CRM_MembershipExtras_DTO_ScheduleInstalmentAmount();
+    $instalment->setAmount($instalmentAmount);
+    $instalment->setTaxAmount($instalmentTaxAmount);
+
+    return $instalment;
+  }
+
+  /**
+   * Calculates the first instalment amount for a set of membership types given that the conditions
+   * for calculating the first instalment amount is met.
+   *
+   * @param CRM_MembershipExtras_DTO_ScheduleInstalmentAmount $instalmentAmount
+   * @param \DateTime|NULL $startDate
+   * @param \DateTime|NULL $endDate
+   * @return CRM_MembershipExtras_DTO_ScheduleInstalment
+   */
+  private function calculateFirstInstalment(
```

Comment:
I don't really understand the point of having this method, I don't see it doing anything special, can you elaborate more on why is it needed?

Reply:
@omarabuhussein  we would need to calculate first instalment different for fixed period membership type.